File: priv/javascript/lib/string.js
import Kernel from './kernel';
import Chars from './string/chars';

function to_atom(string){
  return Symbol.for(string);
}

function to_existing_atom(string){
  return Symbol.for(string);
}

function to_char_list(string){
  return string.split('');
}

function to_float(string){
  return parseFloat(string);
}

function to_integer(string, base = 10){
  return parseInt(string, base);
}

function upcase(binary){
  return binary.toUpperCase();
}

function downcase(binary){
  return binary.toLowerCase();
}

function at(string, position){
  if(position > (string.length - 1)){
    return null;
  }

  return string[position];
}

function capitalize(string){
  let returnString = "";

  for(let i = 0; i < string.length; i++){
    if(i === 0){
      returnString = returnString + string[i].toUpperCase();
    }else{
      returnString = returnString + string[i].toLowerCase();      
    }
  }

  return returnString;
}


function codepoints(string){
  return to_char_list(string).map(function(c){
    return c.codePointAt(0);
  });
}


function contains__qm__(string, contains){
  if(Array.isArray(contains)){
    return contains.some(function(s){
      return string.indexOf(s) > -1;
    });
  }

  return string.indexOf(contains) > -1;
}

function duplicate(subject, n){
  return subject.repeat(n);
}

function ends_with__qm__(string, suffixes){
  if(Array.isArray(suffixes)){
    return suffixes.some(function(s){
      return string.endsWith(s);
    });
  }

  return string.endsWith(suffixes);
}

function first(string){
  if(!string){
    return null;
  }

  return string[0];
}

function graphemes(string){
  return string.split('');
}

function last(string){
  if(!string){
    return null;
  }

  return string[string.length - 1];
}

function length(string){
  return string.length;
}


function match__qm__(string, regex){
  return string.match(regex) != null;
}


function next_codepoint(string){
  if(!string || string === ""){
    return null;
  }

  return Kernel.SpecialForms.tuple(string[0].codePointAt(0), string.substr(1));
}


function next_grapheme(string){
  if(!string || string === ""){
    return null;
  }

  return Kernel.SpecialForms.tuple(string[0], string.substr(1));
}


function reverse(string){
  let returnValue = "";

  for (var i = string.length - 1; i >= 0; i--) {
    returnValue = returnValue + string[i];
  };

  return returnValue;
}


function split(string){
  return string.split();
}


function starts_with__qm__(string, prefixes){
  if(Array.isArray(prefixes)){
    return prefixes.some(function(s){
      return string.startsWith(s);
    });
  }

  return string.startsWith(prefixes);
}

function valid_character__qm__(codepoint){
  try{
    return String.fromCodePoint(codepoint) != null;
  }catch(e){
    return false;
  }
}


export default {
  at,
  capitalize,
  codepoints,
  contains__qm__,
  downcase,
  duplicate,
  ends_with__qm__,
  first,
  graphemes,
  last,
  length,
  match__qm__,
  next_codepoint,
  next_grapheme,
  reverse,
  split,
  starts_with__qm__,
  to_atom,
  to_char_list,
  to_existing_atom,
  to_float,
  to_integer,
  upcase,
  valid_character__qm__,
  Chars
}

File: priv/std_lib/string.ex
defmodule ElixirScript.String do
  @moduledoc false
  import Kernel, except: [length: 1]

  def to_atom(str) do
    JS.Symbol.for(str)
  end

  def to_existing_atom(str) do
    JS.Symbol.for(str)
  end

  def to_char_list(str) do
    str.split("")
  end

  def to_float(str) do
    JS.parseFloat(str)
  end

  def to_integer(str) do
    JS.parseInt(str, 10)
  end

  def to_integer(str, base) do
    JS.parseInt(str, base)
  end

  def upcase(str) do
    str.toUpperCase()
  end

  def downcase(str) do
    str.toLowerCase()
  end

  def at(str, pos) do
    case pos > length(str) do
      true ->
        nil
      _ ->
        str[pos]
    end
  end

  def capitalize(str) do
    first = str[0].toUpperCase()
    rest = str.substr(1).toLowerCase()

    first <> rest
  end

  def split(str) do
    str.split()
  end

  def split(str, replace, options \\ []) do
    limit = Keyword.get(options, :parts, -1)
    trim = Keyword.get(options, :trim, false)
    split = str.split(replace, limit)

    Enum.map(split, fn(x) ->
      if trim do
        x.trim()
      else
        x
      end
    end)
  end


  def next_grapheme(nil), do: nil
  def next_grapheme(""), do: nil

  def next_grapheme(str) do
    { str[0], str.substr(1) }
  end

  def first(nil), do: nil
  def first(str) do
    str[0]
  end

  def last(nil), do: nil
  def last(str) do
    str[length(str) - 1]
  end

  def graphemes(str) do
    str.split('')
  end

  def length(str) do
    str.length()
  end

  def match?(str, regex) do
    str.match(regex) != nil
  end

  def next_codepoint(nil), do: nil
  def next_codepoint(""), do: nil
  def next_codepoint(str) do
    { str[0].codePointAt(0), str.substr(1) }
  end

  def reverse(str) do
    do_reverse(str, "")
  end

  defp do_reverse("", str) do
    str
  end

  defp do_reverse(str, reverse_str) do
    do_reverse(str.substr(1), reverse_str <> last(str))
  end

  def starts_with?(str, prefix) when is_binary(prefix) do
    str.startsWith(prefix)
  end

  def starts_with?(str, prefixes) when is_list(prefixes) do
    do_starts_with?(str, prefixes)
  end

  def do_starts_with?(_, []) do
    false
  end

  def do_starts_with?(str, prefixes) do
    case starts_with?(str, hd(prefixes)) do
      true ->
        true
      _ ->
        do_starts_with?(str, tl(prefixes))
    end
  end


  def ends_with?(str, suffix) when is_binary(suffix) do
    str.endsWith(suffix)
  end

  def ends_with?(str, suffixes) when is_list(suffixes) do
    do_ends_with?(str, suffixes)
  end

  def do_ends_with?(_, []) do
    false
  end

  def do_ends_with?(str, suffixes) do
    case ends_with?(str, hd(suffixes)) do
      true ->
        true
      _ ->
        do_ends_with?(str, tl(suffixes))
    end
  end

  def duplicate(str, n) do
    str.repeat(n)
  end


  def contains?(str, s) when is_binary(s) do
    str.indexOf(s) > -1
  end

  def contains?(str, s) when is_list(s) do
    do_contains?(str, s)
  end

  def do_contains?(_, []) do
    false
  end

  def do_contains?(str, prefixes) do
    case contains?(str, hd(prefixes)) do
      true ->
        true
      _ ->
        do_contains?(str, tl(prefixes))
    end
  end


  def codepoints(str) do
    do_codepoints(str, [])
  end

  def do_codepoints("", codepoint_list) do
    codepoint_list
  end

  def do_codepoints(str, codepoint_list) do
    do_codepoints(str.substr(1), codepoint_list ++ [first(str).codePointAt(0)])
  end

  def valid_character?(codepoint) do
    ElixirScript.Bootstrap.Functions.is_valid_character(codepoint)
  end
end
